File: src/main/java/org/telran/lecture_04_d_and_c/practice/MinMaxFinder/MinMaxFinder.java
package org.telran.lecture_04_d_and_c.practice.MinMaxFinder;

public class MinMaxFinder {
    /**
     * Находит минимальный и максимальный элементы в массиве
     * @param arr массив чисел
     * @return объект, содержащий минимум и максимум
     */
    public static MinMax findMinMax(int[] arr) {
        // Шаг 1: Проверить, что массив не null и не пустой
        // Если массив пустой или null — выбросить исключение IllegalArgumentException

        // Шаг 2: Вызвать вспомогательный рекурсивный метод с аргументами:
        // массив, индекс начала (0), индекс конца (arr.length - 1)

        // Шаг 3: Вернуть результат работы рекурсивного метода
        throw new UnsupportedOperationException("Method findMinMax is not implemented yet");
    }

    /**
     * Рекурсивно находит минимум и максимум в заданном диапазоне массива
     * @param arr массив
     * @param low нижняя граница диапазона (включительно)
     * @param high верхняя граница диапазона (включительно)
     * @return объект MinMax с найденными значениями
     */
    private static MinMax findMinMaxRecursive(int[] arr, int low, int high) {
        // Базовый случай 1: если в диапазоне один элемент
        // Вернуть MinMax, где min = max = этот элемент

        // Базовый случай 2: если в диапазоне два элемента
        // Сравнить их и вернуть MinMax с меньшим и большим значениями

        // Рекурсивный случай: больше двух элементов
        // 1. Найти середину диапазона: mid = low + (high - low) / 2
        // 2. Рекурсивно вызвать метод для левой половины (low, mid)
        // 3. Рекурсивно вызвать метод для правой половины (mid + 1, high)
        // 4. Объединить результаты: найти min из левого и правого min, max из левого и правого max
        // 5. Вернуть новый MinMax с объединёнными значениями

        throw new UnsupportedOperationException("Method findMinMaxRecursive is not implemented yet");
    }

    public static void main(String[] args) {
        int[] arr1 = {3, 5, 1, 8, 9, 2, 6, 7, 4};
        MinMax result1 = findMinMax(arr1);
        System.out.println("Массив: " + arrayToString(arr1));
        System.out.println("Минимум: " + result1.min + ", Максимум: " + result1.max);

        int[] arr2 = {-10, 5, 100, -5, 0, 20, 30};
        MinMax result2 = findMinMax(arr2);
        System.out.println("\nМассив: " + arrayToString(arr2));
        System.out.println("Минимум: " + result2.min + ", Максимум: " + result2.max);

        // Случай с массивом из одного элемента
        int[] arr3 = {42};
        MinMax result3 = findMinMax(arr3);
        System.out.println("\nМассив: " + arrayToString(arr3));
        System.out.println("Минимум: " + result3.min + ", Максимум: " + result3.max);

        // Пустой массив (обработка краевого случая)
        try {
            int[] arr4 = {};
            MinMax result4 = findMinMax(arr4);
            System.out.println("\nМассив: пустой");
            System.out.println("Минимум: " + result4.min + ", Максимум: " + result4.max);
        } catch (IllegalArgumentException e) {
            System.out.println("\nОшибка: " + e.getMessage());
        }
    }



    /**
     * Вспомогательный метод для вывода массива в виде строки
     * @param arr массив
     * @return строковое представление массива
     */
    private static String arrayToString(int[] arr) {
        StringBuilder sb = new StringBuilder("[");
        for (int i = 0; i < arr.length; i++) {
            sb.append(arr[i]);
            if (i < arr.length - 1) {
                sb.append(", ");
            }
        }
        sb.append("]");
        return sb.toString();
    }
}

File: src/main/java/org/telran/lecture_04_d_and_c/practice/MinMaxFinder/min_max_finder.js
class MinMaxFinder {
    /**
     * Находит минимальное и максимальное значение в массиве
     * @param {number[]} arr - массив чисел
     * @returns {{min: number, max: number}} объект с минимумом и максимумом
     */
    static findMinMax(arr) {
        // Шаг 1: Проверить, что массив не пустой и не undefined
        // Если массив пустой — выбросить исключение (например, через throw Error)

        // Шаг 2: Вызвать вспомогательную рекурсивную функцию с индексами начала и конца массива

        // Шаг 3: Вернуть результат работы рекурсивной функции
        return null; // заглушка
    }

    /**
     * Рекурсивно находит минимум и максимум в диапазоне индексов
     * @param {number[]} arr - массив чисел
     * @param {number} low - начальный индекс
     * @param {number} high - конечный индекс
     * @returns {{min: number, max: number}} объект с найденными значениями
     */
    static findMinMaxRecursive(arr, low, high) {
        // Базовый случай 1: один элемент
        // Вернуть объект, где min = max = arr[low]

        // Базовый случай 2: два элемента
        // Сравнить arr[low] и arr[high], вернуть объект с min и max

        // Рекурсивный случай: больше двух элементов
        // 1. Найти середину диапазона: mid = Math.floor((low + high) / 2)
        // 2. Рекурсивно вызвать функцию для левой половины
        // 3. Рекурсивно вызвать функцию для правой половины
        // 4. Объединить результаты: взять min из двух min, max из двух max
        // 5. Вернуть итоговый объект с min и max

        return null; // заглушка
    }

    /**
     * Преобразует массив в строку для вывода
     * @param {number[]} arr
     * @returns {string}
     */
    static arrayToString(arr) {
        // Вернуть строку формата [1, 2, 3] используя join
        return null; // заглушка
    }
}

// --- Тесты ---
const arr1 = [3, 5, 1, 8, 9, 2, 6, 7, 4];
const result1 = MinMaxFinder.findMinMax(arr1);
console.log("Массив:", MinMaxFinder.arrayToString(arr1));
console.log("Минимум:", result1?.min, ", Максимум:", result1?.max);

const arr2 = [-10, 5, 100, -5, 0, 20, 30];
const result2 = MinMaxFinder.findMinMax(arr2);
console.log("\nМассив:", MinMaxFinder.arrayToString(arr2));
console.log("Минимум:", result2?.min, ", Максимум:", result2?.max);

const arr3 = [42];
const result3 = MinMaxFinder.findMinMax(arr3);
console.log("\nМассив:", MinMaxFinder.arrayToString(arr3));
console.log("Минимум:", result3?.min, ", Максимум:", result3?.max);

try {
    const arr4 = [];
    const result4 = MinMaxFinder.findMinMax(arr4);
    console.log("\nМассив: пустой");
    console.log("Минимум:", result4.min, ", Максимум:", result4.max);
} catch (e) {
    console.log("\nОшибка:", e.message);
}
